include/deal.II/grid/manifold_lib.h

```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * optimal mesh quality, those manifolds should be compatible with a chart
+ * notion. For example, computing a point that is 0.25 along the line between
+ * two vertices using the weights 0.25 and 0.75 for the two vertices should
+ * give the same result as first computing the mid point at 0.5 and then again
+ * compute the midpoint between the first vertex and coarse mid point. This is
+ * the case for PolarManifold but not for Spherical manifold, so be careful
+ * when using the latter. In case the quality of the manifold is not good
+ * enough, upon mesh refinement it may happen that the transformation to a
+ * chart inside the get_new_point() or add_new_point() methods produces points
+ * that are outside the unit cells, then this class throws an exception of type
+ * Manifold@<dim,spacedim@>::ExcTransformationFailed. If that happens, the mesh
+ * should be refined before applying this class, as done in the following
+ * example:
+ * @code
+ * SphericalManifold<dim> spherical_manifold;
+ * TransfiniteInterpolationManifold<dim> inner_manifold;
+ * Triangulation<dim> triangulation;
+ * GridGenerator::hyper_ball (triangulation);
+ * triangulation.set_all_manifold_ids(1);
+ * triangulation.set_all_manifold_ids_on_boundary(0);
+ * triangulation.set_manifold (0, polar_manifold);
+ * inner_manifold.initialize(triangulation);
+ * triangulation.set_manifold (1, inner_manifold);
+ * triangulation.refine_global(1);
+ * // initialize the transfinite manifold again
+ * inner_manifold.initialize(triangulation);
+ * triangulation.refine_global(4);
+ * @endcode
+ *
+ * @note For performance and accuracy reasons, it is recommended to apply the
+ * transfinite manifold to as coarse a mesh as possible. Regarding accuracy,
+ * the curved description can only be applied to new points created from a
+ * given neighborhood, and the grid quality is typically higher when extending
+ * the curved description over as large a domain as possible. Regarding
+ * performance, the identification of the correct coarse cell in the
+ * get_new_point() method needs to pass all coarse cells, so expect a quadratic
+ * complexity in the number of coarse cells. Thus, the current implementation
+ * is only really economical when there are not more than a few hundred coarse
+ * cells. To make performance better for larger numbers of cells, one could
+ * extend the current implementation by a pre-identification of relevant cells
+ * with axis-aligned bounding boxes.
+ *
+ * @author Martin Kronbichler, Luca Heltai, 2017
+ */
+template <int dim, int spacedim=dim>
+class TransfiniteInterpolationManifold : public Manifold<dim,spacedim>
+{
+public:
+  /**
+   * Constructor.
+   */
+  TransfiniteInterpolationManifold();
+
+  /**
+   * Initializes the manifold with a coarse mesh. The prerequisite for using
+   * this class is that the input triangulation is uniformly refined and the
+   * manifold is later attached to the same triangulation.
```

Comment:
Maybe we should warn the user that destroying the underlying `Triangulation` will leave this class in an undefined state. Alternatively, we should attach a `boost::signals2::signal<void ()> Triangulation< dim, spacedim >::Signals::clear` to the given triangulation that resets our pointer to be the `nullptr` when the `Triangulation` is cleared.